flink-runtime/src/test/java/org/apache/flink/runtime/executiongraph/AdaptedRestartPipelinedRegionStrategyNGConcurrentFailoverTest.java

```diff
@@ -0,0 +1,282 @@
+
+		final Iterator<ExecutionVertex> vertexIterator = eg.getAllExecutionVertices().iterator();
+		final ExecutionVertex ev11 = vertexIterator.next();
+		final ExecutionVertex ev12 = vertexIterator.next();
+		final ExecutionVertex ev21 = vertexIterator.next();
+		final ExecutionVertex ev22 = vertexIterator.next();
+
+		// start job scheduling
+		eg.scheduleForExecution();
+		manualMainThreadExecutor.triggerAll();
+
+		// fail ev11 to trigger region failover of {ev11}, {ev21}, {ev22}
+		ev11.getCurrentExecutionAttempt().fail(new Exception("task failure"));
+		manualMainThreadExecutor.triggerAll();
+		assertEquals(JobStatus.RUNNING, eg.getState());
+		assertEquals(ExecutionState.FAILED, ev11.getExecutionState());
+		assertEquals(ExecutionState.DEPLOYING, ev12.getExecutionState());
+		assertEquals(ExecutionState.CANCELED, ev21.getExecutionState());
+		assertEquals(ExecutionState.CANCELED, ev22.getExecutionState());
+
+		// trigger global failover cancelling and immediately recovery
+		eg.failGlobal(new Exception("Test global failure"));
+		ev12.getCurrentExecutionAttempt().completeCancelling();
+		manuallyTriggeredRestartStrategy.triggerNextAction();
+		manualMainThreadExecutor.triggerAll();
+
+		// verify the job state and vertex attempt number
+		assertEquals(2, eg.getGlobalModVersion());
+		assertEquals(1, ev11.getCurrentExecutionAttempt().getAttemptNumber());
+		assertEquals(1, ev12.getCurrentExecutionAttempt().getAttemptNumber());
+		assertEquals(1, ev21.getCurrentExecutionAttempt().getAttemptNumber());
+		assertEquals(1, ev22.getCurrentExecutionAttempt().getAttemptNumber());
+
+		// complete region failover blocker to trigger region failover
+		failoverStrategy.getBlockerFuture().complete(null);
+		manualMainThreadExecutor.triggerAll();
+
+		// verify that no task is restarted by region failover
+		assertEquals(ExecutionState.DEPLOYING, ev11.getExecutionState());
+		assertEquals(ExecutionState.DEPLOYING, ev12.getExecutionState());
+		assertEquals(ExecutionState.CREATED, ev21.getExecutionState());
+		assertEquals(ExecutionState.CREATED, ev22.getExecutionState());
+		assertEquals(1, ev11.getCurrentExecutionAttempt().getAttemptNumber());
+		assertEquals(1, ev12.getCurrentExecutionAttempt().getAttemptNumber());
+		assertEquals(1, ev21.getCurrentExecutionAttempt().getAttemptNumber());
+		assertEquals(1, ev22.getCurrentExecutionAttempt().getAttemptNumber());
+	}
+
+	@Test
+	public void testSkipFailoverIfExecutionStateIsNotRunning() throws Exception {
+		final ExecutionGraph executionGraph = createExecutionGraph();
+
+		final Iterator<ExecutionVertex> vertexIterator = executionGraph.getAllExecutionVertices().iterator();
+		final ExecutionVertex firstVertex = vertexIterator.next();
+
+		executionGraph.cancel();
+
+		final FailoverStrategy failoverStrategy = executionGraph.getFailoverStrategy();
+		failoverStrategy.onTaskFailure(firstVertex.getCurrentExecutionAttempt(), new Exception("Test Exception"));
+
```

Comment:
manualMainThreadExecutor.triggerAll() is needed here.

Reply:
Wouldn't hurt but not needed since we are only interested in the state transition to `ExecutionState.CANCELED`.

Reply:
The vertex get directly transitioned to `CANCELED ` because it was no scheduled yet and the graph get cancelled. Seems it does not prove whether the local failover is skipped. (Even though a local failover is not skipped, the recovery action will be queued in main thread executor and not able to change the vertex state out from CANCELED).

If we want to verify "skip local failover if the job status is not Running", how about trigger task failure directly without graph.cancel()? In this case, the local failover will be skipped since the graph is in CREATED state, and the vertex will remain in `CREATED` state.

Without the `triggerAll`, the CANCELED state does not prove that the local failover is skipped, since the local failover may have also triggered the task cancellation and the recovery action is queued in the main thread executor. So the vertex state remains in CANCELLED.

Reply:
I'll add `triggerAll()` since it is a more realistic scenario, and it is less invasive (`createExecutionGraph()` already schedules the `ExecutionGraph`).

Reply:
done